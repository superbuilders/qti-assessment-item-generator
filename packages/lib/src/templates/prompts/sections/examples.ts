import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
