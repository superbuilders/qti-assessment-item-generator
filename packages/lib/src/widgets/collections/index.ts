import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
