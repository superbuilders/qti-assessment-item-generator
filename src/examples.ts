import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
