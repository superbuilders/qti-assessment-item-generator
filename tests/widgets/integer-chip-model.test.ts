import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { integerChipModelExamples } from "../../examples/integer-chip-model"
import type { IntegerChipModelProps } from "../../src/widgets/generators/integer-chip-model"
import type { WidgetInput } from "../../src/widgets/registry"
import { generateWidget } from "../../src/widgets/widget-generator"

describe("Widget: integer-chip-model", () => {
	const examples: WidgetInput[] = integerChipModelExamples

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error, index })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})

	// Additional comprehensive test cases
	describe("edge cases and special scenarios", () => {
		test("should handle single chip", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 300,
				height: 300,
				chips: [{ sign: "plus", crossedOut: false }]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle all crossed-out chips", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 300,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "plus", crossedOut: true }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle narrow width forcing vertical layout", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 300,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle alternating pattern", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 400,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle minimum width constraint", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 300, // Very narrow
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})

	// Database-extracted test cases for comprehensive coverage
	describe("database-extracted test scenarios", () => {
		test("should handle large collection with mixed signs", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle wide layout with many chips", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 300,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle subtraction with single crossed out chip", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: true }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle multiple positive chips crossed out", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: true },
					{ sign: "plus", crossedOut: true },
					{ sign: "plus", crossedOut: true },
					{ sign: "plus", crossedOut: true },
					{ sign: "plus", crossedOut: true },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle mixed crossed out negative chips", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle complex subtraction pattern", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: true },
					{ sign: "plus", crossedOut: true }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle mostly crossed out negative chips", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle large collection with different height", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle complex subtraction with multiple crossed out", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle all negative chips only", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle mixed evaluation scenario", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle small collection with different dimensions", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle single crossed out negative chip", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle mixed crossed out positive chips", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false },
					{ sign: "plus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle partial subtraction pattern", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})

		test("should handle subtraction with crossed out negative chips", async () => {
			const props: IntegerChipModelProps = {
				type: "integerChipModel",
				width: 320,
				height: 300,
				chips: [
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: true },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false },
					{ sign: "minus", crossedOut: false }
				]
			}
			const result = await errors.try(generateWidget(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})
})
