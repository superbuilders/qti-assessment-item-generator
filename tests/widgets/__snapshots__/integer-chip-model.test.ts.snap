// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget: integer-chip-model should produce consistent output for example #1 1`] = `
"<svg width="300" height="150" viewBox="0 0 300 150" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif">
  <circle cx="34" cy="34" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="92" cy="34" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="92" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="150" cy="34" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="150" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="208" cy="34" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="208" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="34" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
</svg>"
`;

exports[`Widget: integer-chip-model should produce consistent output for example #2 1`] = `
"<svg width="320" height="180" viewBox="0 0 320 180" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif">
  <circle cx="34" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="92" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="92" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="150" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="150" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="34" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="92" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="92" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="150" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="150" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
</svg>"
`;

exports[`Widget: integer-chip-model should produce consistent output for example #3 1`] = `
"<svg width="280" height="184" viewBox="0 0 280 184" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif">
  <circle cx="34" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="92" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="92" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="34" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="92" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="92" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="150" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="150" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="208" cy="92" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="208" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="34" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="92" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="92" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="150" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="150" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="208" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="208" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
</svg>"
`;

exports[`Widget: integer-chip-model should produce consistent output for example #4 1`] = `
"<svg width="220" height="242" viewBox="0 0 220 242" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif">
  <circle cx="34" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="92" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="92" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="150" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="150" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="34" cy="92" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="92" cy="92" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="92" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="150" cy="92" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="150" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="34" cy="150" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="92" cy="150" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="92" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="150" cy="150" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="150" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="34" cy="208" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="208" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
</svg>"
`;

exports[`Widget: integer-chip-model should produce consistent output for example #5 1`] = `
"<svg width="360" height="300" viewBox="0 0 360 300" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif">
  <circle cx="34" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="92" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="92" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="150" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="150" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="208" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="208" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="266" cy="34" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="266" y="34" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="34" cy="92" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="34" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="92" cy="92" r="24" fill="#FADBD8" stroke="#212121" stroke-width="2" /><text x="92" y="92" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">−</text>
  <circle cx="34" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="92" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="92" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="150" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="150" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="208" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="208" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="266" cy="150" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="266" y="150" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="34" cy="208" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="208" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="92" cy="208" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="92" y="208" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="150" cy="208" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="150" y="208" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="208" cy="208" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="208" y="208" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="266" cy="208" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="266" y="208" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
  <circle cx="34" cy="266" r="24" fill="#FEF9E7" stroke="#212121" stroke-width="2" /><text x="34" y="266" font-size="32px" font-weight="600" fill="#212121" text-anchor="middle" dominant-baseline="central">+</text>
</svg>"
`;
